axelor-budget/src/main/java/com/axelor/apps/budget/service/BudgetScenarioLineServiceImpl.java

```diff
@@ -70,4 +76,38 @@ public void removeUnusedYears(BudgetScenarioLine budgetScenarioLine, int size) {
       budgetScenarioLine.setYear1Value(BigDecimal.ZERO);
     }
   }
+
+  @Override
+  public List<Map<String, Object>> getLineUsingSection(
+      BudgetLevel section,
+      Set<BudgetScenarioVariable> budgetScenarioVariableSet,
+      List<BudgetScenarioLine> budgetScenarioLineOriginList,
+      List<Map<String, Object>> budgetScenarioLineList) {
+    if (!ObjectUtils.isEmpty(budgetScenarioVariableSet)) {
+      for (BudgetScenarioVariable budgetScenarioVariable : budgetScenarioVariableSet) {
+        BudgetScenarioLine budgetScenarioLine =
+            budgetScenarioLineOriginList.stream()
+                .filter(line -> budgetScenarioVariable.equals(line.getBudgetScenarioVariable()))
+                .findFirst()
+                .orElse(null);
+        if (budgetScenarioLine != null) {
+          BudgetScenarioLine optLine = budgetScenarioLineRepository.copy(budgetScenarioLine, false);
+
+          Map<String, Object> optLineContext = Mapper.toMap(optLine);
+          optLineContext.put("budgetLevel", section.getCode());
+          budgetScenarioLineList.add(optLineContext);
+        }
+      }
+    } else if (section != null && !ObjectUtils.isEmpty(section.getBudgetLevelList())) {
+      for (BudgetLevel child : section.getBudgetLevelList()) {
+        budgetScenarioLineList =
+            getLineUsingSection(
+                child,
+                child.getBudgetScenarioVariableSet(),
+                budgetScenarioLineOriginList,
+                budgetScenarioLineList);
+      }
+    }
+    return budgetScenarioLineList;
+  }
```

Comment:
I don't like the complexity of this method, can you split it and use `while` instead of a recursive method ?